static/js/libs/fileValidation.test.js

```diff
@@ -0,0 +1,321 @@
+import { validateRestrictions } from "./fileValidation";
+
+function generateFile(options) {
+  const fileContents = "testymctestface";
+  return new File([fileContents], "test", options);
+}
+
+describe("validateRestrictions", () => {
+  describe("baseRestrictions", () => {
+    it("should return an error if the file is of the wrong type", async () => {
+      const file = generateFile({
+        type: "text/plain"
+      });
+
+      const accept = ["text/html", "text/rich-as-owt"];
+
+      const validation = await validateRestrictions(file, {
+        accept: accept
+      });
+
+      expect(validation.errors).toEqual([`File type is incorrect`]);
+    });
+
+    it("should return an error if a file is too big", async () => {
+      const file = generateFile({
+        type: "text/html"
+      });
+
+      const max = 0.000014; // 14 bytes
+
+      const validation = await validateRestrictions(file, {
+        size: {
+          max: max
+        }
+      });
+
+      expect(validation.errors).toEqual([`File size is over ${max}MB`]);
+    });
+
+    it("should return an error if a file is too small", async () => {
+      const file = generateFile({
+        type: "text/html"
+      });
+
+      const min = 0.000016; // 16 bytes
+
+      const validation = await validateRestrictions(file, {
+        size: {
+          min: min
+        }
+      });
+
+      expect(validation.errors).toEqual([`File size is below ${min}MB`]);
+    });
+
+    it("should return no errors if the file passes all restrictions", async () => {
+      const file = generateFile({
+        type: "text/html"
+      });
+
+      const restrictions = {
+        accept: ["text/html"],
+        size: {
+          min: 0.000014,
+          max: 0.000016
+        }
+      };
+
+      const validation = await validateRestrictions(file, restrictions);
+
+      expect(validation.errors).toBeUndefined();
+    });
+  });
+
+  describe("imageRestrictions", () => {
+    let restrictions;
+
+    function generateImage(attributes) {
+      class Image {
+        constructor() {
+          this._src = "";
+          this.element = document.createElement("image");
+
+          Object.keys(attributes).forEach(key => {
+            this[key] = attributes[key];
+          });
+        }
+
+        addEventListener(type, callback) {
+          this.element.addEventListener(type, callback);
+        }
+
+        set src(newSrc) {
+          this._src = newSrc;
+          this.element.dispatchEvent(new Event("load"));
+        }
+
+        get src() {
+          return this._src;
+        }
+      }
+      window.Image = Image;
```

Comment:
Not sure how the tests are run in jest, but if we really need to override global objects, we should revert it after test, so it doesn't influence others.

Reply:
I looked this up - because I couldn't find another way to do this... So I found (after digging) in the Jest docs this:
https://jestjs.io/docs/en/configuration#setupfiles-array
> Since every test runs in its own environment

https://jestjs.io/docs/en/configuration#globals-object
> Note that, if you specify a global reference value (like an object or array) here, and some code mutates that value in the midst of running a test, that mutation will not be persisted across test runs for other test files.

And finally
https://jestjs.io/docs/en/configuration#testenvironment-string
> Note: TestEnvironment is sandboxed. Each test suite will trigger setup/teardown in their own TestEnvironment.

From those I took it to mean that each test suite (file) has it's own environment - so as long as that functionality is required in that test file we should be A.OK :+1:

__BUT__ in larger test suites (form.test.js for example) we should definitely revert.

Reply:
OK. It still could be a good practice to just to it with cleanup in before/after callbacks, because any suite can become "larger" at some point and devs adding tests may not be aware that other test case is modifying globals.